pkg/render/fluentd.go

```diff
@@ -427,6 +424,43 @@ func (c *fluentdComponent) envvars() []corev1.EnvVar {
 					},
 				)
 			}
+
+			// When LogTypes is not defined explicitly, the default is to enable
+			// all the log types.
+			if syslog.LogTypes == nil {
```

Comment:
Yeah, that's my bad. 

I added this not knowing the convention for adding defaults and meant to have a comment here to call that out so that we could talk about how that works.

I'll go ahead and do that.

Reply:
I wonder if we can the make the default value of `syslog.LogTypes` to be []syslog.LogTypes{SyslogLogAuditEE,
	SyslogLogAuditKube,
	SyslogLogDNS,
	SyslogLogFlows,
	SyslogLogIDSEvents,}.  What happens if you do not want to export any log to syslog ? How do you disable the feature ?

Reply:
> I wonder if we can the make the default value of syslog.LogTypes to be []syslog.LogTypes{SyslogLogAuditEE,
SyslogLogAuditKube,
SyslogLogDNS,
SyslogLogFlows,
SyslogLogIDSEvents,}

That's an interesting suggestion. You can certainly set default values through the kubebuilder markers. I wonder what the behaviour is though if you have a LogTypes field defined but have it set to empty. 

> What happens if you do not want to export any log to syslog ? How do you disable the feature ?

If you don't want to export any logs of any types to syslog then you should not have the SyslogStoreSpec section at all in your LogCollector CR. 

So if you added it and then decide you don't want to export any logs to syslog, you need to remove the SyslogStoreSpec section entirely. That's the clearest way. 

I think it probably also works if you have the SyslogStoreSpec defined and define an empty LogTypes list. That should work too. But I would treat this second way as unsupported (not documented).

Reply:
I think this it is a bit confusing. I would rather not make this expansion and let the user declare what it needs to forward to syslog.

Reply:
Spoke with @asincu some more. 

It sounds like the design of this field wasn't quite clean enough and it missed some clarity. 

Here's a summary of where we're at: 
- LogTypes is not optional, since if a user has SyslogStoreSpec defined it means they want to export at least one type of log 
- We were trying to compensate for this (in case the user left the field empty) by having it auto-declare the list with all valid log types
- But actually we should make LogTypes required, and make it default to all the valid log types 

We can default it simply be using the `+kubebuilder:default` marker. 

@tmjd thoughts?

Reply:
So the definition will now look like this: 
```yaml
syslog:
  description: If specified, enables exporting of flow, audit, and
    DNS logs to syslog.
  properties:
    endpoint:
      description: 'Location of the syslog server. example: tcp://1.2.3.4:601'
      type: string
    logTypes:
      default:
      - AuditEE
      - AuditKube
      - DNS
      - Flows
      - IDSEvents
      description: LogTypes contains a list of types of logs to
        export to syslog. By default, if this field is omitted,
        it will be set to include all possible values.
      items:
        description: SyslogLogType represents the allowable log
          types for syslog. Allowable values are AuditEE, AuditKube,
          DNS, Flows and IDSEvents. * AuditEE corresponds to audit
          logs for Enterprise API custom resources. * AuditKube
          corresponds to audit logs for Kubernetes API resources.
          * DNS corresponds to DNS logs generated by Calico node.
          * Flows corresponds to flow logs generated by Calico node.
          * IDSEvents corresponds to event logs for the intrusion
          detection system (anomaly detection, suspicious IPs, suspicious
          domains and global alerts).
        enum:
        - AuditEE
        - AuditKube
        - DNS
        - Flows
        - IDSEvents
        type: string
      type: array
    packetSize:
      description: 'PacketSize defines the maximum size of packets
        to send to syslog. In general this is only needed if you
        notice long logs being truncated. Default: 1024'
      format: int32
      type: integer
  required:
  - endpoint
  - logTypes
```

Reply:
I believe setting the Field to required means the user must specify the field when creating the CR. Is that incorrect?
With my understanding, if the user is specifying SyslogSpec then they also have to provide the field. Having it optional allows the user to not specify the field and allows us to default it.

I would say the field is not required because we do not need the user specify information in that field.

If you're confident that the LogCollector.SyslogSpec can be created with no logTypes field and the list will be populated with the list of defaults I'd be ok with this but I do not think that is how it works.

Reply:
I'll test it out against a cluster and confirm what the actual behaviour is.

Reply:
Okay verified that with the field marked as `required` validation prevents it from being empty and it's not automatically populated by the defaults. 

```
# logcollectors.operator.tigera.io "tigera-secure" was not valid:
# * : Invalid value: "The edited file failed validation": ValidationError(LogCollector.spec.additionalStores.syslog): missing required field "logTypes" in io.tigera.operator.v1.LogCollector.spec.additionalStores.syslog
``` 

However, if I set the field back to `optional` **and** I have the default listing defined in the schema it gets auto populated. 

So what I inputed into the LogCollector CR: 
```
spec:
  additionalStores:
    syslog:
      endpoint: tcp://tinysyslog.default.svc:5140
      packetSize: 1024
```

What I got once I saved it: 
```
spec:
  additionalStores:
    syslog:
      endpoint: tcp://tinysyslog.default.svc:5140
      logTypes:
      - AuditEE
      - AuditKube
      - DNS
      - Flows
      - IDSEvents
      packetSize: 1024
```

@tmjd I think that matches up with what are original expectation was. And without needing to write an explicit logic to do that ourselves. 

@asincu are you onboard with this behaviour now? I can see how it was confusing when I was explaining it. 

Here's what we have: 
- If user defines SyslogStoreSpec section and doesn't set logTypes, then it gets automatically populated with the default list (as shown above in this comment)
- If the user decides later on that they don't want to export any logs of any type to syslog, they need to remove the entire SyslogStoreSpec section
- If the user wants to filter down to a smaller subset of log types to export to syslog, then they edit logTypes and remove the types that they don't want to export

👍 👎 ?

Reply:
That sounds reasonable to me.
@caseydavenport is there any reason with this behavior ^ that we need to include the defaulting logic in our code?

I do have a little concern about having the defaulting happening in two different places (I'm mainly thinking it might be confusing when debugging?). For logTypes it happens in the CRD machinery and everything else happening in the controller code. I don't think we could ever move all the defaulting into the CRD either but maybe we can put defaulting like this (where we are just populating a field if the user doesn't specify it) in the CRD and keep "auto-detection" in the controllers.

Reply:
I am a bit on the fence if we need to do this defaulting ourselves.
But both Casey and I have concerns with using the CRD defaulting so I think we should have the logCollector controller handle the defaulting. In addition to the other concerns mentioned, I do not think there will be many fields where we will be able to use the CRD defaulting so I don't want just a few fields defaulted there.

Reply:
To be quite honest, I found the defaulting a bit confusing. I see this more as applying a mathematical function. You provide an empty set or a set with zero elements, you get nothing in return. You provide a collection with only DNS log types, you get only DNS logs. You provide a collection that contains all the log types, you get all the logs in return. I think what we are saying is that if you provide an empty set, you get everything. If you provide a set zero elements you get nothing. 

If the defaulting is being done by the controller, does that mean that the operator needs to make an extra call to update the resource after it was created ?

Reply:
> But both Casey and I have concerns with using the CRD defaulting so I think we should have the logCollector controller handle the defaulting. In addition to the other concerns mentioned, I do not think there will be many fields where we will be able to use the CRD defaulting so I don't want just a few fields defaulted there.

@caseydavenport @tmjd okay I can buy into the reason that fragmenting our defaulting across two different approaches just adds overhead and lacks consistency. 

I'll change it. 

> If the defaulting is being done by the controller, does that mean that the operator needs to make an extra call to update the resource after it was created ?
@asincu from what Erik said it looks like our general pattern does involve an extra update. First we have to grab the CR and examine it, then if a condition is met we update it (here if the logTypes field is empty or not defined but the Syslog section is defined, then populate logTypes with all values). 

In the case of the autobuilder CRD mechanism, I think it might avoid any extra update (perhaps it piggy backs on the original update the user triggered, by running the validation check and then modifying the CR before writing it.

> To be quite honest, I found the defaulting a bit confusing. I see this more as applying a mathematical function. You provide an empty set or a set with zero elements, you get nothing in return. 
@asincu I know it feels a bit awkward from where you standing (i.e. a user would never be able to set logTypes field to empty list). It doesn't let the user do whatever they want with the list, there's a extra mechanism (with default behaviour) happening on top. 

However, if you look at it from a user experience / product perspective it's basically saying the user doesn't have to set logTypes manually (for user convenience/ease of use) and it will auto-populate to having all the types explicitly in the CR (for backwards compatibility if we change the default behaviour in the future).

In the end, if the user wants to not have any logs sent, they _can_ do that, but it requires they remove the Syslog section entirely.

Reply:
Hey @tmjd, the PR is ready for another pass now. I'm hoping to get this merged in the next day or so (before the release branches are cut). 

Here's the [latest changes](https://github.com/tigera/operator/pull/941/commits/4259fbabdb4e4141149098a4f082b1680591b18c).